## Setup
The first thing we typically do when dealing with data in Python is to import the `pandas` package and one of its dependencies, `numpy`. 

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('dhs_service_records_synthesized_final.csv')

## Simple Data Inspection

Before doing anything with the data, it is good practice to take a look at a few samples from the dataset. Here, we're going to take a look at the first ten rows and the last seven rows, or the head and tail of the data; this is a function that is easy to call, and it is good practice to call either of these often so that you know how a dataset evolves over time.

In [3]:
df.head(10)

synthetic_data  MCI_UNIQ_ID  CALDR_YR DATE_OF_EVENT  \
0  SYNTHETIC DATA         1626      2021    01/31/2021   
1  SYNTHETIC DATA         1626      2021    02/28/2021   
2  SYNTHETIC DATA         1626      2021    03/31/2021   
3  SYNTHETIC DATA         1626      2021    04/30/2021   
4  SYNTHETIC DATA         1626      2021    05/31/2021   
5  SYNTHETIC DATA         1626      2021    06/30/2021   
6  SYNTHETIC DATA         3775      2021    01/31/2021   
7  SYNTHETIC DATA         3775      2021    02/28/2021   
8  SYNTHETIC DATA         3775      2021    03/31/2021   
9  SYNTHETIC DATA         3775      2021    04/30/2021   

                                             service  GEO_AREA  age  DOB  DOD  \
0  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
1  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
2  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
3  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
4  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
5  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
6  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
7  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
8  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   
9  Children_Attending_Early_Childhood_Programs_Ma...       NaN    5  NaN  NaN   

   GENDER  GENDER_IDENTITY  SEX_ORIENT  LEGAL_SEX        RACE   ETHNICITY  \
0  1~Male              NaN         NaN        NaN     1~White  99~Unknown   
1  1~Male              NaN         NaN        NaN     1~White  99~Unknown   
2  1~Male              NaN         NaN        NaN     1~White  99~Unknown   
3  1~Male              NaN         NaN        NaN     1~White  99~Unknown   
4  1~Male              NaN         NaN        NaN     1~White  99~Unknown   
5  1~Male              NaN         NaN        NaN     1~White  99~Unknown   
6  1~Male              NaN         NaN        NaN  99~Unknown  99~Unknown   
7  1~Male              NaN         NaN        NaN  99~Unknown  99~Unknown   
8  1~Male              NaN         NaN        NaN  99~Unknown  99~Unknown   
9  1~Male              NaN         NaN        NaN  99~Unknown  99~Unknown   

   LIVING_ARRANGEMENT  EMPLOYMENT_STATUS MARITAL_STATUS  \
0                 NaN                NaN     99~Unknown   
1                 NaN                NaN     99~Unknown   
2                 NaN                NaN     99~Unknown   
3                 NaN                NaN     99~Unknown   
4                 NaN                NaN     99~Unknown   
5                 NaN                NaN     99~Unknown   
6                 NaN                NaN     99~Unknown   
7                 NaN                NaN     99~Unknown   
8                 NaN                NaN     99~Unknown   
9                 NaN                NaN     99~Unknown   

                 EDUCATION_LEVEL  VETERAN_FLAG  
0  9-12~High School (grade 9-12)           NaN  
1  9-12~High School (grade 9-12)           NaN  
2  9-12~High School (grade 9-12)           NaN  
3  9-12~High School (grade 9-12)           NaN  
4  9-12~High School (grade 9-12)           NaN  
5  9-12~High School (grade 9-12)           NaN  
6                     99~Unknown           NaN  
7                     99~Unknown           NaN  
8                     99~Unknown           NaN  
9                     99~Unknown           NaN

In [4]:
df.tail(7)

synthetic_data  MCI_UNIQ_ID  CALDR_YR DATE_OF_EVENT   service  \
7116127  SYNTHETIC DATA       530696      2021    06/30/2021  Suicides   
7116128  SYNTHETIC DATA       530765      2021    06/30/2021  Suicides   
7116129  SYNTHETIC DATA       530832      2021    06/30/2021  Suicides   
7116130  SYNTHETIC DATA       532121      2021    04/30/2021  Suicides   
7116131  SYNTHETIC DATA       533991      2021    11/30/2021  Suicides   
7116132  SYNTHETIC DATA       534127      2021    07/31/2021  Suicides   
7116133  SYNTHETIC DATA       535156      2021    06/30/2021  Suicides   

         GEO_AREA  age  DOB  DOD    GENDER  GENDER_IDENTITY  SEX_ORIENT  \
7116127       NaN   21  NaN  NaN  2~Female              NaN         NaN   
7116128       NaN   15  NaN  NaN  2~Female              NaN         NaN   
7116129       NaN   17  NaN  NaN    1~Male              NaN         NaN   
7116130       NaN   22  NaN  NaN  2~Female              NaN         NaN   
7116131       NaN   16  NaN  NaN    1~Male              NaN         NaN   
7116132       NaN   19  NaN  NaN    1~Male              NaN         NaN   
7116133       NaN   24  NaN  NaN    1~Male              NaN         NaN   

         LEGAL_SEX     RACE              ETHNICITY  LIVING_ARRANGEMENT  \
7116127        NaN  1~White             99~Unknown                 NaN   
7116128        NaN  1~White             99~Unknown                 NaN   
7116129        NaN  1~White             99~Unknown                 NaN   
7116130        NaN  1~White             99~Unknown                 NaN   
7116131        NaN  1~White             99~Unknown                 NaN   
7116132        NaN  1~White  2~Not Hispanic/Latinx                 NaN   
7116133        NaN  1~White             99~Unknown                 NaN   

         EMPLOYMENT_STATUS          MARITAL_STATUS  \
7116127                NaN              99~Unknown   
7116128                NaN              99~Unknown   
7116129                NaN              99~Unknown   
7116130                NaN  1~Single-Never Married   
7116131                NaN  1~Single-Never Married   
7116132                NaN               4~Widowed   
7116133                NaN  1~Single-Never Married   

                       EDUCATION_LEVEL  VETERAN_FLAG  
7116127                     99~Unknown           NaN  
7116128                     99~Unknown           NaN  
7116129                     99~Unknown           NaN  
7116130  9-12~High School (grade 9-12)           NaN  
7116131  9-12~High School (grade 9-12)           NaN  
7116132           GRAD~Graduate Degree           NaN  
7116133  9-12~High School (grade 9-12)           NaN

## Dataframe Attributes 

You can get a lot of helpful information from the attributes of the dataframe; specifically, we are going to look at dataframe dimensions, column names, column data types, and the ranges of column values. To get these, we can easily call some prebuilt attributes that are assigned to pandas dataframes. These help us to understand our data before we even start modifying it.

In [5]:
df.shape

(7116134, 20)

We can see that the dataset has 7116134 rows and 20 columns 

In [6]:
df.columns

Index(['synthetic_data', 'MCI_UNIQ_ID', 'CALDR_YR', 'DATE_OF_EVENT', 'service',
       'GEO_AREA', 'age', 'DOB', 'DOD', 'GENDER', 'GENDER_IDENTITY',
       'SEX_ORIENT', 'LEGAL_SEX', 'RACE', 'ETHNICITY', 'LIVING_ARRANGEMENT',
       'EMPLOYMENT_STATUS', 'MARITAL_STATUS', 'EDUCATION_LEVEL',
       'VETERAN_FLAG'],
      dtype='object')

Pandas dataframes also have a helpful method for easy summarization of the set, which you can call with `df.info()`.

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7116134 entries, 0 to 7116133
Data columns (total 20 columns):
 #   Column              Dtype  
---  ------              -----  
 0   synthetic_data      object 
 1   MCI_UNIQ_ID         int64  
 2   CALDR_YR            int64  
 3   DATE_OF_EVENT       object 
 4   service             object 
 5   GEO_AREA            float64
 6   age                 int64  
 7   DOB                 float64
 8   DOD                 float64
 9   GENDER              object 
 10  GENDER_IDENTITY     float64
 11  SEX_ORIENT          float64
 12  LEGAL_SEX           float64
 13  RACE                object 
 14  ETHNICITY           object 
 15  LIVING_ARRANGEMENT  float64
 16  EMPLOYMENT_STATUS   float64
 17  MARITAL_STATUS      object 
 18  EDUCATION_LEVEL     object 
 19  VETERAN_FLAG        float64
dtypes: float64(9), int64(3), object(8)
memory usage: 1.1+ GB


Another way to check the data types, if you just want that, is to use the `dtypes` attribute of the dataframe.

In [8]:
df.dtypes

synthetic_data         object
MCI_UNIQ_ID             int64
CALDR_YR                int64
DATE_OF_EVENT          object
service                object
GEO_AREA              float64
age                     int64
DOB                   float64
DOD                   float64
GENDER                 object
GENDER_IDENTITY       float64
SEX_ORIENT            float64
LEGAL_SEX             float64
RACE                   object
ETHNICITY              object
LIVING_ARRANGEMENT    float64
EMPLOYMENT_STATUS     float64
MARITAL_STATUS         object
EDUCATION_LEVEL        object
VETERAN_FLAG          float64
dtype: object

## Taking a look at some records
 - take a look at 3 random rows, it is important to inspect data beyond the head and tail

In [9]:
# set seed for random function so that we get same rows when re-run the cell
np.random.seed(17)
np.random.randint(0, df.shape[0],3)

array([ 589041, 5539983, 2078086])

In [10]:
np.random.randint(3, df.shape[0],3) # gives different three numbers

array([ 491929,  828732, 6417139])

In [11]:
# set seed for random function so that we get same rows when re-run the cell
np.random.seed(17)
df.iloc[np.random.randint(0, df.shape[0],3)]

synthetic_data  MCI_UNIQ_ID  CALDR_YR DATE_OF_EVENT  \
589041   SYNTHETIC DATA       364757      2021    10/31/2021   
5539983  SYNTHETIC DATA       449747      2021    09/30/2021   
2078086  SYNTHETIC DATA       109525      2021    09/30/2021   

                                       service  GEO_AREA  age  DOB  DOD  \
589041      Individuals_Receiving_DHS_Services       NaN   54  NaN  NaN   
5539983  Individuals_Receiving_Income_Supports       NaN   51  NaN  NaN   
2078086  Individuals_Receiving_Income_Supports       NaN   51  NaN  NaN   

         GENDER  GENDER_IDENTITY  SEX_ORIENT  LEGAL_SEX     RACE  \
589041   1~Male              NaN         NaN        NaN  1~White   
5539983  1~Male              NaN         NaN        NaN  1~White   
2078086  1~Male              NaN         NaN        NaN  1~White   

                     ETHNICITY  LIVING_ARRANGEMENT  EMPLOYMENT_STATUS  \
589041   2~Not Hispanic/Latinx                 NaN                NaN   
5539983  2~Not Hispanic/Latinx                 NaN                NaN   
2078086  2~Not Hispanic/Latinx                 NaN                NaN   

                 MARITAL_STATUS EDUCATION_LEVEL  VETERAN_FLAG  
589041   1~Single-Never Married      99~Unknown           NaN  
5539983  1~Single-Never Married      99~Unknown           NaN  
2078086  1~Single-Never Married      99~Unknown           NaN

 - take a look at the record for a specific MCI_UNIQ_ID, say merchant 364757 from above result

In [12]:
df[df['MCI_UNIQ_ID'] == 364757].sort_values("DATE_OF_EVENT")

synthetic_data  MCI_UNIQ_ID  CALDR_YR DATE_OF_EVENT  \
6710438  SYNTHETIC DATA       364757      2021    01/31/2021   
4675581  SYNTHETIC DATA       364757      2021    01/31/2021   
589034   SYNTHETIC DATA       364757      2021    02/28/2021   
4675582  SYNTHETIC DATA       364757      2021    02/28/2021   
589035   SYNTHETIC DATA       364757      2021    03/31/2021   
6710439  SYNTHETIC DATA       364757      2021    03/31/2021   
4675583  SYNTHETIC DATA       364757      2021    03/31/2021   
6710440  SYNTHETIC DATA       364757      2021    04/30/2021   
4675584  SYNTHETIC DATA       364757      2021    04/30/2021   
589036   SYNTHETIC DATA       364757      2021    05/31/2021   
6710441  SYNTHETIC DATA       364757      2021    05/31/2021   
4675585  SYNTHETIC DATA       364757      2021    05/31/2021   
6710442  SYNTHETIC DATA       364757      2021    06/30/2021   
4675586  SYNTHETIC DATA       364757      2021    06/30/2021   
589037   SYNTHETIC DATA       364757      2021    06/30/2021   
4675587  SYNTHETIC DATA       364757      2021    07/31/2021   
6710443  SYNTHETIC DATA       364757      2021    07/31/2021   
589038   SYNTHETIC DATA       364757      2021    07/31/2021   
4675588  SYNTHETIC DATA       364757      2021    08/31/2021   
6710444  SYNTHETIC DATA       364757      2021    08/31/2021   
589039   SYNTHETIC DATA       364757      2021    08/31/2021   
4675589  SYNTHETIC DATA       364757      2021    09/30/2021   
589040   SYNTHETIC DATA       364757      2021    09/30/2021   
589041   SYNTHETIC DATA       364757      2021    10/31/2021   
6710445  SYNTHETIC DATA       364757      2021    10/31/2021   
4675590  SYNTHETIC DATA       364757      2021    10/31/2021   
589042   SYNTHETIC DATA       364757      2021    11/30/2021   
4675591  SYNTHETIC DATA       364757      2021    11/30/2021   
6710446  SYNTHETIC DATA       364757      2021    11/30/2021   
4675592  SYNTHETIC DATA       364757      2021    12/31/2021   
589043   SYNTHETIC DATA       364757      2021    12/31/2021   
6710447  SYNTHETIC DATA       364757      2021    12/31/2021   

                                              service  GEO_AREA  age  DOB  \
6710438  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
4675581         Individuals_Receiving_Income_Supports       NaN   54  NaN   
589034             Individuals_Receiving_DHS_Services       NaN   54  NaN   
4675582         Individuals_Receiving_Income_Supports       NaN   54  NaN   
589035             Individuals_Receiving_DHS_Services       NaN   54  NaN   
6710439  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
4675583         Individuals_Receiving_Income_Supports       NaN   54  NaN   
6710440  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
4675584         Individuals_Receiving_Income_Supports       NaN   54  NaN   
589036             Individuals_Receiving_DHS_Services       NaN   54  NaN   
6710441  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
4675585         Individuals_Receiving_Income_Supports       NaN   54  NaN   
6710442  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
4675586         Individuals_Receiving_Income_Supports       NaN   54  NaN   
589037             Individuals_Receiving_DHS_Services       NaN   54  NaN   
4675587         Individuals_Receiving_Income_Supports       NaN   54  NaN   
6710443  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
589038             Individuals_Receiving_DHS_Services       NaN   54  NaN   
4675588         Individuals_Receiving_Income_Supports       NaN   54  NaN   
6710444  Individuals_Receiving_Mental_Health_Services       NaN   54  NaN   
589039             Individuals_Receiving_DHS_Services       NaN   54  NaN   
4675589         Individuals_Receiving_Income_Supports       NaN   54  NaN   
589040             Individuals_Receiving_DHS_Services       NaN   54  NaN   
589041             Individuals_Receiving_DHS_Ser